jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -1566,18 +1566,21 @@ private void completeChannelCallback(Throwable failure)
                 stream = httpChannelState._stream;
                 assert httpChannelState._callbackFailure == null;
 
-                // Turn pending demand or unconsumed input on persistent connections into failure
+                // Turn pending demand into failure.
                 if (httpChannelState._onContentAvailable != null)
+                {
                     failure = ExceptionUtil.combine(failure, new IllegalStateException("demand pending"));
-                else if (httpChannelState.getConnectionMetaData().isPersistent())
-                    failure = ExceptionUtil.combine(failure, stream.consumeAvailable());
+                }
                 else
                 {
+                    // consumeAvailable() makes the connection non-persistent and returns an exception
+                    // when the content cannot be consumed, this must not result in an error according
+                    // to RFC2616 section 8.2.3.
                     Throwable unconsumed = stream.consumeAvailable();
```

Comment:
Can you explain why we need to consumeAvailable on a connection that is already not persistent?